src/control/lib/control/pool_test.go

```diff
@@ -1244,6 +1282,94 @@ func TestPoolGetProp(t *testing.T) {
 			}
+
+			// Verify response can be marshalled without error.
+			_, err := json.Marshal(gotResp)
+			if err != nil {
+				t.Fatalf("Unexpected error: %s\n", err.Error())
+			}
+		})
+	}
+}
+
+func TestControl_PoolGetPropResp_MarshalJSON(t *testing.T) {
+	for name, tc := range map[string]struct {
+		resp    []*daos.PoolProperty
+		expData string
+		expErr  error
+	}{
+		"nil": {
+			expData: "null",
+		},
+		"all props": {
+			resp: []*daos.PoolProperty{
+				propWithVal("checkpoint", "timed"),
+				propWithVal("checkpoint_freq", "10000"),
+				propWithVal("checkpoint_thresh", "20"),
+				propWithVal("ec_cell_sz", "4096"),
+				propWithVal("ec_pda", "1"),
+				propWithVal("global_version", "1"),
+				propWithVal("label", "foo"),
+				propWithVal("perf_domain", "root"),
+				propWithVal("policy", "type=io_size"),
+				propWithVal("rd_fac", "1"),
+				propWithVal("reclaim", "disabled"),
+				propWithVal("rp_pda", "2"),
+				propWithVal("scrub", "timed"),
+				propWithVal("scrub-freq", "1024"),
+				propWithVal("scrub-thresh", "0"),
+				propWithVal("self_heal", "exclude"),
+				propWithVal("space_rb", "42"),
+				func() *daos.PoolProperty {
+					p := propWithVal("svc_list", "")
+					p.Value.SetString("[0-3]")
+					return p
+				}(),
+				propWithVal("svc_rf", "3"),
+				propWithVal("upgrade_status", "in progress"),
+			},
+			expData: `[{"name":"checkpoint","description":"WAL Checkpointing behavior","value":"timed"},{"name":"checkpoint_freq","description":"WAL Checkpointing frequency, in seconds","value":10000},{"name":"checkpoint_thresh","description":"Usage of WAL before checkpoint is triggered, as a percentage","value":20},{"name":"ec_cell_sz","description":"EC cell size","value":4096},{"name":"ec_pda","description":"Performance domain affinity level of EC","value":1},{"name":"global_version","description":"Global Version","value":1},{"name":"label","description":"Pool label","value":"foo"},{"name":"perf_domain","description":"Pool performance domain","value":"root"},{"name":"policy","description":"Tier placement policy","value":"type=io_size"},{"name":"rd_fac","description":"Pool redundancy factor","value":1},{"name":"reclaim","description":"Reclaim strategy","value":"disabled"},{"name":"rp_pda","description":"Performance domain affinity level of RP","value":2},{"name":"scrub","description":"Checksum scrubbing mode","value":"timed"},{"name":"scrub-freq","description":"Checksum scrubbing frequency","value":1024},{"name":"scrub-thresh","description":"Checksum scrubbing threshold","value":0},{"name":"self_heal","description":"Self-healing policy","value":"exclude"},{"name":"space_rb","description":"Rebuild space ratio","value":42},{"name":"svc_list","description":"Pool service replica list","value":[0,1,2,3]},{"name":"svc_rf","description":"Pool service redundancy factor","value":3},{"name":"upgrade_status","description":"Upgrade Status","value":"in progress"}]`,
+		},
+		"missing props; v2_2 pool": {
```

Comment:
I'm a little confused by this test case. Is the error expected because in the normal case, the properties with no value should be filtered out from the response?

Reply:
yes this is the error that caused the issue we are trying to solve, this verifies that Unmarshal doesn't allow missing values for existing properties